containers/daap-python-base/src/var/task/versioning.py

```diff
@@ -202,35 +203,85 @@ def update_metadata_remove_schemas(self, schema_list: list[str]) -> str:
+        )
+        self._copy_from_previous_version(
+            latest_version=self.latest_version, new_version=new_version
+        )
+        new_version_key = self.data_product_config.metadata_path(new_version).key
+        metadata.write_json_to_s3(new_version_key)
+
+        return new_version
+
+    def _verify_input_metadata(self, input_data: dict) -> DataProductMetadata:
+        """Load and verify input Data Product metadata, raising errors if metadata is
+        missing, invalid, or if there are changes to fields that cannot be altered
+        between versions.
+
+        Args:
+            input_data (dict): input metadata in a format ready for loading into a
+            DataProductMetadata object.
+
+        Raises:
+            InvalidUpdate: raised if metadata is missing, invalid, or if there are
+            changes to fields that cannot be altered between versions
+
+        Returns:
+            DataProductMetadata: verified metadata object with current metadata in
+            metadata.data and previous metadata in metadata.latest_version_saved_data
         """
         metadata = DataProductMetadata(
             data_product_name=self.data_product_config.name,
             logger=self.logger,
             input_data=input_data,
         ).load()
+
         if not metadata.valid or not metadata.exists:
-            raise InvalidUpdate()
+            msg = "invalid metadata passed"
+            raise InvalidUpdate(msg)
+
+        changed_fields = metadata.changed_fields()
+        if changed_fields.difference(UPDATABLE_METADATA_FIELDS):
+            changed_fields = list(changed_fields)
+            num_fields = len(changed_fields)
+            msg = f"Non-updatable metadata field{('s'[:num_fields!=1])} changed:"
+            for f in changed_fields:
+                msg += f"{f}: {metadata.latest_version_saved_data[f]} -> {metadata.data[f]}; "
+            self.logger.error(msg)
+            raise InvalidUpdate(msg)
+
+        return metadata
+
+    def update_metadata(self, input_data) -> str:
+        """
+        Create a new version with updated metadata.
+        """
+        metadata = self._verify_input_metadata(input_data)
 
         state = metadata_update_type(metadata)
         self.logger.info(f"Update type {state}")
 
         if state != UpdateType.MinorUpdate:
             raise InvalidUpdate(state)
 
-        latest_version = metadata.version
-        new_version = generate_next_version_string(latest_version, state)
-        self._copy_from_previous_version(
-            latest_version=latest_version, new_version=new_version
-        )
-        new_version_key = self.data_product_config.metadata_path(new_version).key
-        metadata.write_json_to_s3(new_version_key)
+        new_version = self._minor_version_bump(metadata)
 
         return new_version
 
-    def update_schema(self, input_data: dict, table_name: str):
+    def update_schema(self, input_data: dict, table_name: str) -> tuple[str, dict]:
```

Comment:
Yeah this felt a little uncomfortable: it was based on retaining the logic of there being a pre-existing `DataProductSchema` object in the lambda handler already.

For the `create_schema.handler` an existing `schema` object is used for:
- determining whether to commit to the `version_manager.create_schema` call
    - `schema.exists`, `schema.has_registered_data_product`, `schema.valid`
- pushing the schema to the catalogue

`version_manager.create_schema` could instead return a `DataProductSchema` object:
- this object gets passed to push to catalogue
- we could handle the 3 schema checks inside the `version_manager.create_schema` method

Reply:
Yep, either that or the other way round - we could change all the other methods to accept objects instead of dictionaries. But I think I prefer the above.

Reply:
Done